### PR TITLE
Add `annotateSourceFile` option

### PR DESCRIPTION
**Pull request by natemoo-re**
## Changes

- Allows Astro compiler to annotate elements with information about their source file and location
- Planned for improving DX in Astro core during dev

## Testing

- Tests added

## Docs

N/A

### CODE REVIEW

File: internal/transform/transform_test.go [hunk 1]
```diff
@@ -199,3 +199,47 @@ func TestFullTransform(t *testing.T) {
...
+		},
+		{
+			name:   "no components",
+			source: `<Component>Hello world!</Component>`,
+			want:   `<Component>Hello world!</Component>`,
+		},
+		{
+			name:   "injects root",
+			source: `<html></html>`,
+			want:   `<html data-astro-source-root="file://"></html>`,
```
Comment (natemoo-re) [inline internal/transform/transform_test.go:222]: We inject `data-astro-source-root={transformOpts.ProjectRoot}` to the `html` element

File: internal/transform/transform.go [hunk 2]
```diff
@@ -11,15 +11,16 @@ import (
...
+	Scope              string
+	Filename           string
+	Pathname           string
+	InternalURL        string
+	SourceMap          string
+	Site               string
+	ProjectRoot        string
+	PreprocessStyle    interface{}
+	StaticExtraction   bool
+	AnnotateSourceFile bool
```
Comment (natemoo-re) [inline internal/transform/transform.go:23]: This is a new `annotateSourceFile: boolean` option for the compiler

**Comment by Princesseuh** [comment]
Tested it with https://github.com/withastro/astro/pull/9016, and it works!

**Review by natemoo-re — COMMENTED** [review]
Approved!

File: internal/printer/print-to-js.go [hunk 3]
```diff
@@ -448,6 +449,25 @@ func render1(p *printer, n *Node, opts RenderOptions) {
 				// Note: if we encounter "slot" NOT inside a component, that's fine
 				// These should be preserved in the output
 				p.printAttribute(a, n)
+			} else if a.Key == "data-astro-source-file" {
+				p.printAttribute(a, n)
+				var l []int
```
Comment (ematipico) [inline internal/printer/print-to-js.go:454]: What the hell is `l`?? 😆 

L anime

**Review by ematipico — APPROVED** [review]

File: internal/printer/print-to-js.go (hunk 3 above)
Comment (Princesseuh) [reply internal/printer/print-to-js.go:454]: an average variable in the compiler, ha. All the variables follow this naming scheme

Comment (natemoo-re) [reply internal/printer/print-to-js.go:454]: Yeah this aligns with the style of this codebase (which was adapted from the existing style of Go's official HTML parser), but I absolutely agree that one letter variables suck. They're all over the codebase and it makes it very difficult to read at a glance.

**Pull request merged**